addons/survey/models/survey.py

```diff
@@ -0,0 +1,1077 @@
+        # Let's get back, baby!
+        elif go_back and survey.users_can_go_back:
+            return (pages[current_page_index - 1][1], current_page_index - 1, False)
+        else:
+            # This will show the last page
+            if current_page_index == len(pages) - 2:
+                return (pages[current_page_index + 1][1], current_page_index + 1, True)
+            # This will show a regular page
+            else:
+                return (pages[current_page_index + 1][1], current_page_index + 1, False)
+
+    def filter_input_ids(self, survey, filters, finished=False):
+        '''If user applies any filters, then this function returns list of
+           filtered user_input_id and label's strings for display data in web.
+           :param filters: list of dictionary (having: row_id, ansewr_id)
+           :param finished: True for completely filled survey,Falser otherwise.
+           :returns list of filtered user_input_ids.
+        '''
+        if filters:
+            domain_filter, choice = [], []
+            for filter_data in filters:
+                row_id, answer_id = filter_data['row_id'], filter_data['answer_id']
+                if row_id == 0:
+                    choice.append(answer_id)
+                else:
+                    domain_filter.extend(['|', ('value_suggested_row.id', '=', row_id), ('value_suggested_id.id', '=', answer_id)])
+            if choice:
+                domain_filter.insert(0, ('value_suggested_id.id', 'in', choice))
+            else:
+                domain_filter = domain_filter[1:]
+            user_input_lines = self.env['survey.user_input_line'].search(domain_filter)
+            filtered_input_ids = user_input_lines.mapped('user_input_id').ids if user_input_lines else []
+        else:
+            filtered_input_ids = []
+        if finished:
+            UserInput = self.env['survey.user_input']
+            if not filtered_input_ids:
+                user_inputs = UserInput.search([('survey_id', '=', survey.id)])
+            else:
+                user_inputs = UserInput.browse(filtered_input_ids)
+            return user_inputs.filtered(lambda user_input: user_input.state == 'done').ids
+        return filtered_input_ids
+
+    @api.model
+    def get_filter_display_data(self, filters):
+        '''Returns data to display current filters
+        :param filters: list of dictionary (having: row_id, answer_id)
+        :param finished: True for completely filled survey, False otherwise.
+        :returns list of dict having data to display filters.
+        '''
+        filter_display_data = []
+        if filters:
+            Surveylabel = self.env['survey.label']
+            for filter_data in filters:
+                row_id, answer_id = filter_data['row_id'], filter_data['answer_id']
+                question = Surveylabel.browse(answer_id).question_id
+                if row_id == 0:
+                    labels = Surveylabel.browse([answer_id])
```

Comment:
you are already browsing at line 352, do it something like,
``` survey_label = Surveylabel.browse(answer_id) ```
Use it at line 354 and at line 357 question_text = survey_label.question_id.question